### DISCUSSION TITLE
從撕書來看電腦怎麼search

### DISCUSSION THREAD

**Discussion by Tomlord1122** [1101-GS4719-C/Python-Community-GS4719-C#213] [General]
哈佛有一個很有名的課叫做CS50，有點像是我們的通識課，但是他以非常硬為名，用一整個學期來教大家寫程式。
youtube上就有教學影片了。

其中有一部分讓我印象很深刻，台上的教授用了一本厚厚的原文書示範了甚麼叫以電腦的角度來作search。
他不斷的把書給撕成兩半，直到他找到了要找到的頁數為止，這個概念有點像是我們在玩終極密碼，假如數字範圍是1-1000，答
案是125好了，第一次我猜500，猜錯一次以後，我們就會知道答案是在1-500裡面，藉由這個方法，我們可以把搜尋125這個數字所花的時間大大的減少，以電腦來說，他每次都會從數字1開始找，這樣如果要找125等於他花了125次才找到正確的答案。以時間複雜度來說就是O(N)，但用剛剛的撕書的角度來看每一回合我們都是以n/2這樣下去找，時間複雜度就是O(logN)，這種二分法，我們就可以引入資料結構非常重要的一個概念，二元樹(binary tree)，就是不斷的向左or向右的一棵樹。
![下載](https://user-images.githubusercontent.com/79390871/146957456-dc39d1fe-9ee0-49d7-9aeb-57396e7b94dc.png)

**Comment by Tomlord1122** [comment]
二元樹應用的範圍非常的廣，有延伸出很多不同的排序法，不同的樹怎麼應用，把樹高給減少等等。

**Comment by luo910227** [comment]
有找到二元樹的優缺點：
優點是容易實作，且能快速找到任意節點的父節點與左右子節點。 缺點是當二元樹稀疏或不平衡時，就會相當浪費記憶體。
我了解到二元樹在記憶體足夠且撰寫適當的話，的確能省下不少的程式執行時間，感謝你的介紹。
https://kopu.chat/2017/06/18/%E4%BA%8C%E5%85%83%E6%A8%B9/

**Comment by jcpan821** [comment]
感謝分享~~
